elab-util-extra/tests/specialise-data/_shared/Shared.idr

```diff
@@ -0,0 +1,172 @@
+  Elab ()
+verifyEq polyTy specTy lhs rhs = do
+  lhsNorm <- normaliseAs specTy lhs
+  rhsNorm <- normaliseAs specTy rhs
+  lhsCast <- normaliseAs polyTy `(cast ~lhsNorm)
+  rhsCast <- normaliseAs polyTy `(cast ~rhsNorm)
+  checkSoEq `(~lhsNorm == ~rhsNorm) `(~lhsCast == ~rhsCast)
+  checkSoEq `(~lhsNorm /= ~rhsNorm) `(~lhsCast /= ~rhsCast)
+
+||| Display the results of running Show implementations for polymorphicType and specialisedType
+||| on a given value
+verifyShow : (polymorphicType, specialisedType : Type) -> (specialisedValue : TTImp) -> Elab ()
+verifyShow polyTy specTy val = do
+  valNorm <- normaliseAs specTy val
+  valCast <- normaliseAs polyTy `(cast ~valNorm)
+  s1 <- normaliseAs String `(show ~valNorm)
+  s2 <- normaliseAs String `(show ~valCast)
+  checkSoEq s1 s2
+
+
+||| Run polymorphic and monomorphic `show` for every value in list
+verifyShows :
+  (polymorphicType, specialisedType : Type) -> (specialisedValues: List TTImp) -> Elab ()
+verifyShows polyTy specTy = traverse_ $ verifyShow polyTy specTy
+
+||| Verify correctness of DecEq for every possible pair of values from a given list
+verifyDecEqs :
+  (polymorphicType, specialisedType : Type) -> (specialisedValues: List TTImp) -> Elab ()
+verifyDecEqs polyTy specTy vals = do
+  for_ [| MkPair vals vals |] $ uncurry $ verifyDecEq polyTy specTy
+
+
+||| Verify correctness of Eq for every possible pair of values from a given list
+verifyEqs :
+  (polymorphicType, specialisedType : Type) -> (specialisedValues: List TTImp) -> Elab ()
+verifyEqs polyTy specTy vals = do
+  for_ [| MkPair vals vals |] $ uncurry $ verifyEq polyTy specTy
+
+||| Verify specialisation of polymorphicType into specialisedType for a given list of value pairs
+||| where the first element is the value of the polymorphic type, and the second - of the monomorphic one
+export
+verifySpecialisation' :
+  (polymorphicType, specialisedType : Type) -> (valuePairs: List (TTImp, TTImp)) -> Elab ()
+verifySpecialisation' polyTy specTy pairs = do
+  let (polyVals, specVals) = unzip pairs
+  constructExprs specTy specVals
+  verifySingleCasts' polyTy specTy pairs
+  verifyDoubleCasts polyTy specTy specVals
+  verifyDoubleCasts specTy polyTy polyVals
+  case (!(search $ DecEq polyTy), !(search $ DecEq specTy)) of
+    (Just _, Just _) => verifyDecEqs polyTy specTy specVals
+    (Nothing, Just _) => fail "Specialised type implements DecEq, while polymorhpic type doesn't"
+    (Just _, Nothing) => fail "Polymorphic type implements DecEq, while specialised type doesn't"
+    (Nothing, Nothing) => pure ()
+  case (!(search $ Show polyTy), !(search $ Show specTy)) of
+    (Just _, Just _) => verifyShows polyTy specTy specVals
+    (Nothing, Just _) => fail "Specialised type implements Show, while polymorhpic type doesn't"
+    (Just _, Nothing) => fail "Polymorphic type implements Show, while specialised type doesn't"
+    (Nothing, Nothing) => pure ()
+  case (!(search $ Eq polyTy), !(search $ Show specTy)) of
```

Comment:
Inconsistent, should be `Eq` on both sides.

I recommend merging all these checks in something like
```idris
for_ [`(DecEq), `(Show), `(Eq)] $ \expr => do case (!(search $ ~expr polyTy), ...) of ...
```